index.js

```diff
@@ -1,82 +1,41 @@
 'use strict';
 
-const url = require('url');
 const rule = require('unified-lint-rule');
 const visit = require('unist-util-visit');
-const linkCheck = require('link-check');
-const isRelativeUrl = require('is-relative-url');
+const checkLinks = require('check-links');
 const isOnline = require('is-online');
 
-const defaultCache = {};
-const pending = new Map();
-const protocolWhitelist = new Set(['https:', 'http:']);
-
 function noDeadUrls(ast, file, options) {
-  let baseUrl;
-  if (typeof options === 'string') {
-    baseUrl = options;
-    options = {};
-  } else {
-    options = options || {};
-    baseUrl = options.baseUrl;
-  }
-  const cache = options.cache || defaultCache;
+  const urlToNodes = {};
 
-  const warn = node => {
-    file.message(`Link to ${node.url} is dead`, node);
-  };
+  const aggregate = node => {
+    const url = node.url;
+    if (!url) return;
 
-  const skipUrl = subjectUrl => {
-    if (isRelativeUrl(subjectUrl)) {
-      return !baseUrl;
-    } else {
-      const parsedUrl = url.parse(subjectUrl);
-      return !protocolWhitelist.has(parsedUrl.protocol);
+    if (!urlToNodes[url]) {
+      urlToNodes[url] = [];
     }
+
+    urlToNodes[url].push(node);
   };
 
-  // Ensure there's at least one Promise to resolve
-  const promises = [Promise.resolve()];
-  const validate = node => {
-    let subjectUrl = node.url;
-    if (!subjectUrl) return;
+  visit(ast, 'link', aggregate);
+  visit(ast, 'definition', aggregate);
+  visit(ast, 'image', aggregate);
 
-    if (skipUrl(subjectUrl)) return;
+  return checkLinks(Object.keys(urlToNodes), options).then(results => {
+    for (const url in results) {
```

Comment:
Should we include a `hasOwnProperty` check if we're going to use `for...in`?

Reply:
It's not necessary if the results are a plain JS object. My understanding is that `for...in` and `for...of` are preferred over other `for` and `forEach` approaches, though tbh it doesn't really matter to me as it's all evaluates to roughly the same thing during JIT.

Happy to change this to your preferred for variant if you'd like :)

Reply:
I'd prefer an `Object.keys(x).forEach(y)`, for the sake of following a universally safe pattern.